commands/src/main/java/org/wildfly/extras/creaper/commands/security/AddLoginModule.java

```diff
@@ -0,0 +1,167 @@
+import org.wildfly.extras.creaper.core.online.CliException;
+import org.wildfly.extras.creaper.core.online.OnlineCommand;
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.OperationException;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.Values;
+
+/**
+ * Add a new authentication classic login module to given security domain.
+ */
+public class AddLoginModule implements OnlineCommand, OfflineCommand {
+
+    private final String securityDomainName;
+    private final String name;
+    private final String code;
+    private final String flag;
+    private final String module;
+    private final Map<String, String> moduleOptions;
+    private final boolean replaceExisting;
+
+    protected AddLoginModule(Builder builder) {
+        this.securityDomainName = builder.securityDomainName;
+        this.name = builder.name;
+        this.code = builder.code;
+        this.flag = builder.flag;
+        this.module = builder.module;
+        this.moduleOptions = builder.moduleOptions;
+        this.replaceExisting = builder.replaceExisting;
+    }
+
+    @Override
+    public final void apply(OnlineCommandContext occ) throws CliException, CommandFailedException, IOException,
+            TimeoutException, InterruptedException {
+        Operations ops = new Operations(occ.client);
+        Address authenticationClassicAddress = Address.subsystem("security")
+                .and("security-domain", securityDomainName)
+                .and("authentication", "classic");
+        try {
+            boolean exists = ops.exists(authenticationClassicAddress);
+            if (!exists) {
+                ops.add(authenticationClassicAddress);
+            }
+        } catch (OperationException e) {
+            throw new IOException("Failed to access or create authentication=classic in security domain "
+                    + securityDomainName, e);
+        }
+
+        Address loginModuleAddress = authenticationClassicAddress.and("login-module", name);
+
+        if (replaceExisting) {
+            try {
+                ops.removeIfExists(loginModuleAddress);
+            } catch (OperationException e) {
+                throw new IOException("Failed to remove existing login module " + name + " in security domain "
+                        + securityDomainName, e);
+            }
+        }
+
+        ops.add(loginModuleAddress, Values.empty()
+                .andOptional("code", code)
+                .andOptional("flag", flag)
+                .andOptional("module", module)
+                .andObjectOptional("module-options", Values.fromMap(moduleOptions))
+        );
+    }
+
+    @Override
+    public final void apply(OfflineCommandContext occ) throws CommandFailedException, IOException {
```

Comment:
I wonder why do you call it `occ` :-) Not that it's bad, but I was hoping to establish a convention of `ctx`.